src/common/tensor_inspector.h

```diff
@@ -0,0 +1,717 @@
+    }
+    if (tb_.ndim() == 1) {
+       dict += ",";
+    }
+    dict += ")} ";
+    int padding_size = 64 - ((10 + dict.size()) % 64);
+    dict += std::string(padding_size, ' ');
+    dict[dict.size()-1] = '\n';
+    std::string header;
+    header += static_cast<char>(0x93);
+    header += "NUMPY";
+    header += static_cast<char>(0x01);
+    header += static_cast<char>(0x00);
+    header += static_cast<char>((uint16_t)dict.size() & 0x00ff);
+    header += static_cast<char>(((uint16_t)dict.size() >> 8) & 0x00ff);
+    header += dict;
+    InspectorManager::get()->dump_value_tag_counter_[tag] += 1;
+    int visit = InspectorManager::get()->dump_value_tag_counter_[tag];
+    std::ofstream file(tag + "_" + std::to_string(visit) + ".npy",
+        std::ios::out | std::ios::binary);
+    file.write(header.c_str(), header.size());
+    file.write(reinterpret_cast<char*>(tb_.dptr<DType>()), sizeof(DType) * tb_.shape_.Size());
+    file.close();
+  }
+
+  /* !\brief the tensor blob */
+  const TBlob tb_;
+  /* !\brief the run context of the tensor */
+  const RunContext& ctx_;
+
+ public:
+   /*!
+   * \brief construct from Tensor object
+   * \tparam Device the device the tensor resides in
+   * \tparam dimension the dimension of the tensor
+   * \tparam DType the data type
+   * \param ts the source tensor object
+   * \param ctx the run context of the tensor
+   */
+  template<typename Device, int dimension,
+      typename DType MSHADOW_DEFAULT_DTYPE>
+  TensorInspector(const mshadow::Tensor<Device, dimension, DType>& ts, const RunContext& ctx):
+      tb_(ts), ctx_(ctx) {}
+
+  /*!
+   * \brief construct from TBlob object
+   * \param tb the source tblob object
+   * \param ctx the run context of the tensor
+   */
+  TensorInspector(const TBlob& tb, const RunContext& ctx):
+      tb_(tb), ctx_(ctx) {}
+
+  /*!
+   * \brief construct from NDArray object. Currently this only works with kDefaultStorage
+   * \param arr the source ndarray object
+   * \param ctx the run context of the tensor
+   */
+  TensorInspector(const NDArray& arr, const RunContext& ctx):
+      tb_(arr.data()), ctx_(ctx) {}
+
+  /*!
+   * \brief print the tensor to std::cout
+   */
+  inline void print_string() {
+    std::cout << to_string() << std::endl;
+  }
+
+  /*!
+   * \brief return a string which contains the values and other info of the tensor
+   */
+  inline std::string to_string() {
+    std::stringstream ss;
+    MSHADOW_TYPE_SWITCH(tb_.type_flag_, DType, {
+      to_string_helper<DType>(&ss);
+    });
+    return ss.str();
+  }
+
+  /*!
+   * \brief interactively print the tensor value
+   * \param tag the name given to this call
+   */
+  inline void interactive_print(std::string tag = "") {
+    MSHADOW_TYPE_SWITCH(tb_.type_flag_, DType, {
+      interactive_print_helper<DType>(tag);
+    });
+  }
+
+  /*!
+   * \brief check/validate the values within the tensor, return the coordinates
+   * where the lambda evaluates to true
```

Comment:
can you elaborate which lambda you are referring?

Reply:
I meant "return the coordinates where the value checker evaluates to true". Will update the comment